### DISCUSSION TITLE
Some (emacs) global keys are inaccessible in insert state but accessible in normal state?

### DISCUSSION THREAD

**Discussion by Inc0n** [meow-edit/meow#255] [Q&A] [answered]
The global keys I am refering to is such as the basic: 
- `M-f`, `M-b`,
- `C-M-i` (the completion), `C-M-k` (kill-sexp), `M-DEL`, 
- including `M-x`?
 
Which are all accessible in normal state, but inaccessible in insert state.

There are some others which as accessible, such as:
- C-f, C-b

Is this the intended behaviour? Could it be a mis-configuration on my side? 

From the above non-exhausstive listing, it seems something is affecting the meta key for these binds to function properly under insert state.

**Comment by DogLooksGood** [comment]
By default `m` means `M-`, and `g` means `C-M-`. So `M-f` is `SPC m f`, and `C-M-f` is `SPC g f`.

The prefixs are customizable.

**Reply by Inc0n** [reply]
It's great to hear that theses could be customized, however, I still don't see how I can access `M-f` in insert state?
Could you point me into the right direciton please, to provide some examples that would bring more clarity.
Thank you

**Reply by DogLooksGood** [reply]
You just use `M-f` in insert state. Meow provides a modal based on vanilla Emacs keybinding, you can still use all builtins.

**Reply by Inc0n** [reply]
It's great to hear that but as you can see in my descriptions I am having trouble accessing these however.
Would you happen to have any idea that why this might be?

**Reply by DogLooksGood** [reply]
Check what command is bound to `M-f`?

**Reply by Inc0n** [reply]
It is undefined. 
My apology, I have came to realised that this is because I have bound `C-[` to `meow-insert-exit` in insert state, which caused all meta prefixed keys inaccessible. 
This was possible to do in evil.

**Comment by Inc0n** [comment, marked as answer]
I have bound `C-[` to `meow-insert-exit` in insert state which made all meta prefixed keys inaccessible.
Which is a habit carried over from `evil` of some sort.

**Reply by Linerre** [reply]
I used to be a Vim user and learned that `C-[` means `ESC` (to exit `INSERT`). 
Since I switched to Emacs and meow, I have never bound `C-[` to anything in any of meow's modes and it behaves exactly like `ESC`.

Proof:
1. `C-h k` to describe a keybinding
2. `C-[`
3. `ESC- ` (Emacs automatically translates `C-[` to `ESC` and thinks I want to use it as a prefix'.

So I agree with you on that is not a good habit.

**Reply by Inc0n** [reply]
There is a post I found about this topic on [stackoverflow](https://emacs.stackexchange.com/questions/7832/how-to-bind-c-for-real) about it and in more details.
Essentially, the work around I found that if you really desire `C-[` binds is to have:
```
(define-key input-decode-map [?\C-\[] [C-\[])
(define-key global-map [C-\[] [?\C-g])
```
And replace `[?\C-g]` with your configuration of choice, but of course, as outlined in the post (link) above, this may have undesired consequences when using emacs in terminals. Since I am a GUI user, I am satisfied with this.

Note: `(kbd "C-[")` and `[?\C-\[]` are equivalent.
And for better readability, you may want to use `[control-bracketleft]` also.

**Reply by loganrios** [reply]
+1 to this--I had the same idea (mapping `C-[` to `meow-insert-exit`) and had been struggling to figure out why all my `M-<key>` bindings had broken. Thanks so much for documenting your solution!
